### PR TITLE
Feature:Prevent segmentation fault & update new example:syscall_hijack

### PR DESCRIPTION
**Pull request by Sy0307**
## Description

Fixes # (issue)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

- [ ] Test A
- [ ] Test B

**Test Configuration**:

- Firmware version:
- Hardware:
- Toolchain:
- SDK:

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: example/syscall_hijack/open_modify.bpf.c [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+struct {
+	__uint(type, BPF_MAP_TYPE_RINGBUF);
+	__uint(max_entries, 256 * 1024);
+} rb SEC(".maps");
+
+SEC("tracepoint/syscalls/sys_enter_open")
+int tracepoint__syscalls__sys_enter_open(struct trace_event_raw_sys_enter *ctx)
+{
+	// spdlog::info("tracepoint__syscalls__sys_enter_open");
```
Comment (yunwei37) [inline example/syscall_hijack/open_modify.bpf.c:31]: remove these comments

File: example/syscall_hijack/dummy.c [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+int main(int argc, char **argv)
+{
+    // 等待一段时间，以便外部完成对eBPF map的PID注入更新
+    int pid = getpid();
+    printf("current pid: %d\n", pid);
+    // sleep(20);
+    printf("pending until input\n");
+    int c = getchar(); 
+
+    // 设置要执行的victim程序路径
```
Comment (yunwei37) [inline example/syscall_hijack/dummy.c:13]: All comments should be in English

File: runtime/src/bpf_helper.cpp [hunk 3]
```diff
@@ -64,20 +66,92 @@ long bpftime_strncmp(const char *s1, uint64_t s1_sz, const char *s2)
...
+		perror("sigaction");
+		return -EFAULT;
+	}
+	
+	if (sigsetjmp(jump_buffer, 1) == 0) {
+        // to avoid faulting for src & dst (which is NULL)	
+		memcpy((void *)(uintptr_t)dst, (void *)(uintptr_t)ptr,
 	       (size_t)(uint32_t)(size));
-	return 0;
+		spdlog::debug("probe_read: dst={}, src={}, len={}", dst, ptr, size);
```
Comment (yunwei37) [inline runtime/src/bpf_helper.cpp:100]: Use trace log here

File: runtime/src/bpf_helper.cpp [hunk 4]
```diff
@@ -64,20 +66,92 @@ long bpftime_strncmp(const char *s1, uint64_t s1_sz, const char *s2)
...
-	       (size_t)(uint32_t)(len));
-	return 0;
+	if(len<=0) {
+		return -EINVAL;
+	}
+    struct sigaction sa, old_sa;
+    uint64_t ret = 0;
+
+    sa.sa_handler = segv_handler; // set signal handler
+    sigemptyset(&sa.sa_mask); // clear signal set
```
Comment (yunwei37) [inline runtime/src/bpf_helper.cpp:129]: Can we further optimize the code here?

**Review by yunwei37 — CHANGES_REQUESTED** [review]
1. Please split the example and fix into 2 different pr.
2. All comments should be in English
3. Avoid comment code without any explanation.
4. Please add more tests if possible.

File: runtime/src/bpf_helper.cpp [hunk 5]
```diff
@@ -64,20 +66,92 @@ long bpftime_strncmp(const char *s1, uint64_t s1_sz, const char *s2)
...
+{
+    siglongjmp(jump_buffer, 1);
+}
+
+
 uint64_t bpftime_probe_read(uint64_t dst, uint64_t size, uint64_t ptr, uint64_t,
 			    uint64_t)
 {
-	memcpy((void *)(uintptr_t)dst, (void *)(uintptr_t)ptr,
+	if(size<=0) {
```
Comment (Officeyutong) [inline runtime/src/bpf_helper.cpp:81]: Please format your code with vendored clang format file

File: runtime/src/bpf_helper.cpp [hunk 6]
```diff
@@ -64,20 +66,92 @@ long bpftime_strncmp(const char *s1, uint64_t s1_sz, const char *s2)
...
+	}
+	
+	if (sigsetjmp(jump_buffer, 1) == 0) {
+        // to avoid faulting for src & dst (which is NULL)	
+		memcpy((void *)(uintptr_t)dst, (void *)(uintptr_t)ptr,
 	       (size_t)(uint32_t)(size));
-	return 0;
+		spdlog::debug("probe_read: dst={}, src={}, len={}", dst, ptr, size);
+    } else {
+		spdlog::error("probe_read: failed to read from src={}", ptr);
```
Comment (Officeyutong) [inline runtime/src/bpf_helper.cpp:102]: Please use macros to log, so we can control log level at compile time. For example, SPDLOG_ERROR(...)

File: runtime/src/bpf_helper.cpp [hunk 7]
```diff
@@ -64,20 +66,92 @@ long bpftime_strncmp(const char *s1, uint64_t s1_sz, const char *s2)
...
+	}
+    struct sigaction sa, old_sa;
+    uint64_t ret = 0;
+
+    sa.sa_handler = segv_handler; // set signal handler
+    sigemptyset(&sa.sa_mask); // clear signal set
+    sa.sa_flags = 0; 
+
+    if (sigaction(SIGSEGV, &sa, &old_sa) == -1) {
+        perror("sigaction");
```
Comment (Officeyutong) [inline runtime/src/bpf_helper.cpp:133]: We should print errors in logs, not to console

**Review by Officeyutong — CHANGES_REQUESTED** [review]

File: runtime/src/bpf_helper.cpp (hunk 6 above)
Comment (Sy0307) [reply runtime/src/bpf_helper.cpp:102]: @yunwei37 suggested we use SPDLOG_TRACE here, I modify it by SPDLOG_TRACE. Thanks!

Comment (yunwei37) [reply runtime/src/bpf_helper.cpp:102]: No, for error here, you should use SPDLOG_ERROR.

For debug purposes, change to SPDLOG_DEBUG or SPDLOG_TRACE

Comment (Sy0307) [reply runtime/src/bpf_helper.cpp:102]: I got it, thanks.

**Comment by Sy0307** [comment]
As comment below, I will split it into two individual PRs.

**Pull request closed without merging**